specifyweb/frontend/js_src/lib/components/Statistics/index.tsx

```diff
@@ -0,0 +1,599 @@
+        categories: Object.entries(pageStatsSpec).map(
+          ([categoryName, { label, items }]) => ({
+            label,
+            items: Object.entries(items ?? {}).map(([itemName]) => ({
+              type: 'DefaultStat',
+              pageName: pageName,
+              categoryName,
+              itemName,
+            })),
+          })
+        ),
+      })),
+    [statsSpec]
+  );
+}
+
+function useCustomStatQueryBuilderSpec(queryId: number):
+  | {
+      readonly tableName: keyof Tables;
+      readonly fields: RA<
+        Partial<SerializedResource<SpQueryField>> & { readonly path: string }
+      >;
+      readonly label: string;
+    }
+  | undefined {
+  const [data] = useAsyncState(
+    React.useCallback(
+      async () =>
+        fetchResource('SpQuery', queryId).then((queryData) => ({
+          tableName: queryData.contextName as keyof Tables,
+          fields: queryData.fields.map((field) => ({
+            ...field,
+            path: QueryFieldSpec.fromStringId(
+              field.stringId,
+              field.isRelFld ?? false
+            )
+              .toMappingPath()
+              .join('.'),
+          })),
+          label: queryData.name,
+        })),
+      [queryId]
+    ),
+    false
+  );
+  return data;
+}
+
+function DefaultStatItem({
+  statsSpec,
+  pageName,
+  categoryName,
+  itemName,
+  onRemove: handleRemove,
+  onClick: handleClick,
+}: {
+  readonly statsSpec: IR<
+    IR<{
+      readonly label: string;
+      readonly items: StatCategoryReturn;
+    }>
+  >;
+  readonly pageName: string;
+  readonly categoryName: keyof typeof statsSpec;
+  readonly itemName: string;
+  readonly onRemove: (() => void) | undefined;
+  readonly onClick: (() => void) | undefined;
+}): JSX.Element {
+  const statSpecItemPage = statsSpec[pageName];
+  const statSpecItemObject = statSpecItemPage[categoryName]?.items;
+  const statSpecItem =
+    statSpecItemObject === undefined ? undefined : statSpecItemObject[itemName];
+  const statValue =
+    statSpecItem === undefined ? undefined : statSpecItem.spec.type ===
+      'Querybuildstat' ? (
+      <QueryBuilderStat
+        fields={statSpecItem.spec.fields}
+        statLabel={statSpecItem.label}
+        tableName={statSpecItem.spec.tableName}
+        onClick={handleClick}
+        onRemove={handleRemove}
+      />
+    ) : (
+      <StatsResult
+        query={undefined}
+        statLabel={statSpecItem?.label}
+        statValue={statSpecItem.spec.value}
+        onClick={handleClick}
+        onRemove={handleRemove}
+      />
+    );
+  return statValue ?? <p> {commonText('loading')}</p>;
+}
+
+function CustomStatItem({
+  queryId,
+  onRemove: handleRemove,
+  onClick: handleClick,
+}: {
+  readonly queryId: number;
+  readonly onRemove: (() => void) | undefined;
+  readonly onClick: (() => void) | undefined;
+}): JSX.Element {
+  const { tableName, fields, label } =
+    useCustomStatQueryBuilderSpec(queryId) ?? {};
+  const statValue =
```

Comment:
Same as for previous one. Replace `const statValue = ` with `return` and `undefined` with `<p>{commonText('loading')}</p>`